// @meiraba/discord/utils/Constants.js
module.exports.opCode = {
    DISPATCH: 0,
    HEARTBEAT: 1,
    IDENTIFY: 2,
    PRESENCE_UPDATE: 3,
    VOICE_STATE_UPDATE: 4,
    RESUME: 6,
    RECONNECT: 7,
    RESUEST_GUILD_MEMBERS: 8,
    INVALID_SESSION: 9,
    HELLO: 10,
    HEARTBEAT_ACK: 11,
}
module.exports.opCode_voice = {
    IDENTIFY: 0,
    SELECT_PROTOCOL: 1,
    READY: 2,
    HEARTBEAT: 3,
    SESSION_DESCRIPTION: 4,
    SPEAKING: 5,
    HEARTBEAT_ACK: 6,
    RESUME: 7,
    HELLO: 8,
    RESUMED: 9,
    CLIENT_DESCONNECT: 13
}
module.exports.properties = {
    os: "windows",
    browser: "discord-ws",
    device: "discord-ws",
}

module.exports.intents = {
    GUILDS: 1 << 0,

    MEMBERS: 1 << 1,

    BANS: 1 << 2,

    EMOJIS_AND_STICKERS: 1 << 3,

    INTEGRATIONS: 1 << 4,

    WEBHOOKS: 1 << 5,

    INVITES: 1 << 6,

    VOICE_STATES: 1 << 7,

    PRESENCES: 1 << 8,

    MESSAGES: 1 << 9,

    MESSAGE_REACTIONS: 1 << 10,

    MESSAGE_TYPING: 1 << 11,

    DIRECT_MESSAGES: 1 << 12,

    DIRECT_MESSAGE_REACTIONS: 1 << 13,

    DIRECT_MESSAGE_TYPING: 1 << 14,

    MESSAGE_CONTENT: 1 << 15,

    SCHEDULED_EVENTS: 1 << 16,

    AUTO_MODERATION_CONFIGURATION: 1 << 20,

    AUTO_MODERATION_EXECUTION: 1 << 21,
}
